text/0020-network-protocol-pubsub.md

```diff
@@ -0,0 +1,321 @@
+
+# Summary
+
+[summary]: #summary
+
+Tremor Messaging Protocol Specification - Publish/Subscribe Model
+
+# Motivation
+
+[motivation]: #motivation
+
+Extends the tremor network protocol specification with publish/subscribe
+message delivery semantics
+
+Enables publishing data to tremor and subscribing to data from tremor
+over the tremor network protocol.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+This specification normatively defines the `pubsub` protocol extension
+of the tremor network protocol specification core.
+
+A `pubsub` session can be established by the [connect] control plane
+event message with an optional list of initial subscriptions to data
+sources available to tremor.
+
+Once subscribed, events received on subscriptions will be forwarded
+to the `pubsub` session.
+
+Subscriptions are tremor urls that encapsulate data sources. It SHOULD
+be possible to subscribe to onramp or source connectors as well as the
+default or named user defined output streams of tremor pipelines.
+
+To publish data to pipeline inputs or directly to the inputs of data
+sink or offramp connectors requires a subscription for publication.
+
+This specification reserves the string literal `pubsub` as a valid built in
+protocol type supported by implementations of the tremor network protocol
+specification.
+
+It is an error to attempt to register a protocol backend that
+consumes the `pubsub` literal or that overrides the implementation provided
+by tremor itself.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+This section describes how to `connect`, `subscribe` and `unsubscribe`
+from data sources available within tremor over the tremor network protocol.
+
+This section further describes the `list` utility function and the correct
+behaviour upon a normal `close` operation on an established `pubsub` session.
+
+The `pubsub` extension allows publish-subscribe message channels to
+be established via control plane `connect` messages.
+
+## Connect
+
+A minimally correct instanciation is as follows:
+
+```json
+  { "tremor": 
+    { "connect": 
+      { "protocol": "pubsub",
+      }
+    }
+  }
+```
+
+Connecting with an initial subscription that taps into a
+source connector deployed within the connected tremor instance:
+
+```json
+  { "tremor": 
+    { "connect": 
+      { "protocol": "pubsub",
+        "properties": {
+          "subscriptions": [
+            "/onramps/metronome/in"
+          ]
+        }
+      }
+    }
+  }
+```
+
+Connecting with an initial subscription into a pipeline deployed
+within the connected tremor instance:
+
+```json
+  { "tremor": 
+    { "connect": 
+      { "protocol": "pubsub",
+        "properties": {
+          "subscriptions": [
+            "/pipeline/passthrough/out"
+          ]
+        }
+      }
+    }
+  }
+```
+
+Connecting with an initial subscription for publication that
```

Comment:
I think if we support publishing, we need to clarify how events originating from a pubsub session are identified in terms of eventid and origin uri. if a publications session is just forwarding events from somewhere else that already have some identifier, we are good. Should we require events to already be identified or do we want to introduce ids and origin uris for pubsub sessions?
 This might be a future thing to add.

Reply:
Yes, there's going to be some iteration in the specification here as `pipeline` and `connector` endpoints differ significantly and this may move us to a slightly more verbose mapping ( exposing more metadata ). Specifically, tapping into sources and sinks may force handling this more explicitly. As we progress the implementation spike we can true up this draft accordingly